ghproxy/ghcache/coalesce_test.go

```diff
@@ -92,37 +107,144 @@ func TestRoundTrip(t *testing.T) {
+			checkMode(resp, ModeMiss)
 		}
-		resp, err := rt.RoundTrip(req)
-		if err != nil {
-			res <- err
-		} else if b, err := ioutil.ReadAll(resp.Body); err != nil {
-			res <- err
-		} else if string(b) != "Response" {
-			res <- errors.New("unexpected response value")
+		wg.Done()
+	}()
+	// There is a race here and where sleeps are used below.
+	// We need to wait for the initial request to be made
+	// to the coalescer before letting upstream respond, but we don't have a way
+	// of knowing when the requests has actually started waiting on the
+	// responseWaiter...
+	time.Sleep(time.Second * 3)
+
+	// Queue a second request for resource1.
+	// This should coalesce and eventually return ModeCoalesced.
+	wg.Add(1)
+	go func() {
+		if resp, err := runRequest(coalesce, "/resource1", false); err != nil {
+			t.Errorf("Failed to run request: %v.", err)
+		} else {
+			checkMode(resp, ModeCoalesced)
 		}
-		res <- nil
+		wg.Done()
+	}()
+	time.Sleep(time.Second * 3)
+
+	// Requests should be waiting now. Start responding and wait for all
+	// downstream responses to return.
+	delegate.beginResponding.Broadcast()
+	wg.Wait()
+
+	// A later request for resource1 revalidates cached response.
+	// This should return ModeRevalidated.
+	header := http.Header{}
+	header.Set("Status", "304 Not Modified")
+	delegate.responseHeader = header
+	if resp, err := runRequest(coalesce, "/resource1", true); err != nil {
+		t.Errorf("Failed to run request: %v.", err)
+	} else {
+		checkMode(resp, ModeRevalidated)
+	}
+
+	// Another request for resource1 after the resource has changed.
+	// This should return ModeChanged.
+	header = http.Header{}
+	header.Set("X-Conditional-Request", "I am an E-Tag.")
+	delegate.responseHeader = header
+	if resp, err := runRequest(coalesce, "/resource1", true); err != nil {
+		t.Errorf("Failed to run request: %v.", err)
+	} else {
+		checkMode(resp, ModeChanged)
 	}
-	go run()
+
+	// Request for new resource2 with no concurrent requests.
+	// This should return ModeMiss.
+	delegate.responseHeader = nil
+	if resp, err := runRequest(coalesce, "/resource2", true); err != nil {
+		t.Errorf("Failed to run request: %v.", err)
+	} else {
+		checkMode(resp, ModeMiss)
+	}
+
+	// Request for uncacheable resource3.
+	// This should return ModeNoStore.
+	header = http.Header{}
+	header.Set("Cache-Control", "no-store")
+	delegate.responseHeader = header
+	if resp, err := runRequest(coalesce, "/resource3", true); err != nil {
+		t.Errorf("Failed to run request: %v.", err)
+	} else {
+		checkMode(resp, ModeNoStore)
+	}
+
+	// We never send a ModeError mode in a header because we never return a
+	// http.Response if there is an error. ModeError is only for metrics.
+
+	// Mind as well validate hit count in this test too.
+	expectedHits := map[string]int{"/resource1": 3, "/resource2": 1, "/resource3": 1}
+	if !reflect.DeepEqual(delegate.hits, expectedHits) {
+		t.Errorf("Unexpected hit count(s). Expected %v, but got %v.", expectedHits, delegate.hits)
```

Comment:
Can we do `ObjectReflectDiff`? It has nice output

Reply:
Done.